### PR TITLE
[API] Implement custom type adapter for Date

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*
#221 

*Description of changes:*
~~Created custom parser for deserializing/serializing `Date` class in GSON.~~

**_core_**

Four *new* data types added:
* `AWSTemporal` <- Abstract class implemented by the next three
* `AWSDate`
* `AWSTime`
* `AWSDateTime`

All three types wrap around a `GregorianCalendar` object, storing a specific `TimeZone` instance as well as unix epoch time (obtainable as either `long` or `java.util.Date`).

**_aws-api_**

`AWSTemporalTypeAdapter` class added to *aws-api* package, where it is registered as a custom type adapter for any instance of `AWSTemporal` when de/serializing `GraphQLResponse`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    private static final TimeZone DEFAULT_TIME_ZONE = TimeZone.getDefault();
+
+    private static Calendar cal;
+    private static Gson gson;
+
+    /**
+     * Configure gson to use the correct custom type adapter for Date class.
+     */
+    @BeforeClass
+    public static void setup() {
```
Comment (jamesonwilliams) [inline aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java]: By default, unit tests should not use static fields or methods. We use `@BeforeClass` in our tests only because of the design of `Amplify`, as a static singleton. But, this type of thing is usually considered a code smell.

TLDR: strip `static` from this file and switch to `@Before`.

File: aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+     * Test that a string containing just date can be parsed to Date.
+     *
+     * Amplify will assume local timezone if none is specified. This
+     * means that time (in UNIX epoch) is ACTUALLY being affected by
+     * the local timezone of the machine that is parsing the string.
+     */
+    @Test
+    public void testDateWithoutOffset() {
+        cal = new GregorianCalendar(DEFAULT_TIME_ZONE);
+        resetFields(cal);
```
Comment (jamesonwilliams) [inline aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java]: Can you do [`cal.setTimeInMillis(0)`](https://docs.oracle.com/javase/7/docs/api/java/util/Calendar.html#setTimeInMillis(long))?

File: aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+//        String serialized = gson.toJson(date);
+//        /*
+//         * They cannot be exactly same because date will
+//         * pick up local timezone if none is specified.
+//         * Local timezone is dependent on the machine
+//         * this test is running on.
+//         *
+//         * Should be "2020-01-11-08:00" if running from PST.
+//         */
+//        assertEquals(json, serialized);
```
Comment (jamesonwilliams) [inline aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java]: You may be able to correct this issue by adding an overloaded constructor into `AWSDateTimeTypeAdapter`. e.g.,

```
final class AWSDateTimeTypeAdapter {
    AWSDateTimeTypeAdapter(Dependency dependency) {
    }

    AWSDateTimeTypeAdapter() {
       this(Dependendency.defaultInstance());
    }
}
```

In the current code, I think `Dependency` is a `Calendar`, actually? And you'd have to snake it into your util class, too? Or actually, maybe what you need is the ability to pass in a `TimeZone` as the dependency, and then use the current platform time zone as the default, but in test, pass in an arranged time zone. This way, your tests always work in an expected way.

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSTemporalTypeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Factory to be registered with GSON when dealing with AWS AppSync
+ * scalars. This adapter will correctly de/serialize strings that are
+ * correctly formatted according to the specifications for AWSDate,
+ * AWSTime, and AWSDateTime.
+ *
+ * @see <a href="https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html">AWS AppSync Defined Scalars</a>
+ */
+@SuppressWarnings("ParameterName") // Staying faithful to original method parameter names
+public final class AWSTemporalTypeAdapter implements TypeAdapterFactory {
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSTemporalTypeAdapter.java]: If you generic-ize this a bit, you can probably end up with something like:
```
// <T> is AWSDateTime, AWSDate, AWSTime.
final class NullSafeParseAdapter<T> implements TypeAdapterFactory {
    ParseMethod<T> method;
    NullSafeParseAdapter(ParseMethod<T> method) {
        this.method = method;
    }
   @Override
   public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
       // Same as what you have now, except with type <T> 
       // And in the read() method, call return method.apply(in.nextString())
       // Instead of hard-coding a call to your util
   }
   interface ParseMethod<T> {
       T parse(String string);
   }
}

gson
    .registerTypeAdapterFactory(AWSDateTime.class, new NullSafeParseAdapter(AWSDateTime::fromString))
    .registerTypeAdapterFactory(AWSDate.class, new NullSafeParseAdapter(AWSDate::fromString))
    .registerTypeAdapterFactory(AWSTime.class, new NullSafeParseAdapter(AWSTime::fromString))
```

File: aws-api/src/main/java/com/amplifyframework/api/aws/GsonGraphQLResponseFactory.java [hunk 5]
```diff
@@ -17,6 +17,9 @@
 
 import com.amplifyframework.AmplifyException;
 import com.amplifyframework.api.ApiException;
+import com.amplifyframework.api.aws.internal.AWSDateTimeTypeAdapter;
+import com.amplifyframework.api.aws.internal.AWSDateTypeAdapter;
+import com.amplifyframework.api.aws.internal.AWSTimeTypeAdapter;
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/GsonGraphQLResponseFactory.java]: renamed package to `scalar`. `internal` was a terrible name since it will be exposed to people

File: core/src/main/java/com/amplifyframework/core/model/types/JavaFieldType.java [hunk 6]
```diff
@@ -50,15 +50,20 @@
...
-     * Represents the java.util.Date data type.
+     * Represents the java.sql.Date data type.
      */
-    DATE(java.util.Date.class.getSimpleName()),
+    DATE(java.sql.Date.class.getSimpleName()),
 
     /**
      * Represents the java.sql.Time data type.
      */
     TIME(java.sql.Time.class.getSimpleName()),
```
[the remark this answers is not in the document]
Comment (raphkim) [reply core/src/main/java/com/amplifyframework/core/model/types/JavaFieldType.java]: replaced all with custom types

File: aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java]: yep

File: aws-api/src/main/java/com/amplifyframework/api/aws/internal/AWSDateTimeUtils.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    private static final int MINUTES_IN_ONE_HOUR = 60;
+    private static final int SECONDS_IN_ONE_HOUR = 60;
+    private static final int MILLIS_IN_ONE_SECOND = 1000;
+
+    private static final String DATE_REGEX_EXPRESSION =
+            "(?:([0-9]{4})-(1[0-2]|[0][1-9])-(3[01]|[12][0-9]|0[1-9]))";
+    private static final String TIME_REGEX_EXPRESSION =
+            "(?:(2[0-3]|[01][0-9]):([0-5][0-9])(?::([0-5][0-9])(?:.([0-9]{3}))?)?)";
+    private static final String ZONE_REGEX_EXPRESSION =
+            "(?:Z|([+-])(2[0-3]|[01][0-9])(?::([0-5][0-9])(?::([0-5][0-9]))?)?)";
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/internal/AWSDateTimeUtils.java]: scrapped regex

File: aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java (hunk 2 above)
Comment (raphkim) [reply aws-api/src/test/java/com/amplifyframework/api/aws/GsonAWSDateTimeTest.java]: im dumb. you are absolutely right

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDate.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        String timezone = AWSDateTimeUtils.format(getTimeZone());
+        return String.format(Locale.US, "%04d-%02d-%02d%s",
+                getYear(),
+                getMonth() + 1,
+                getDayOfMonth(),
+                timezone);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDate.java]: both `hashcode()` + `equals()` are now `final`

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTime.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ * DateTime instance to accommodate the specifications of
+ * AppSync Scalar AWSDateTime data type.
+ *
+ * Format: YYYY-MM-DDThh:mm:ss.sss[Z|±hh:mm:ss]
+ *
+ * Note: Timezone offset is compulsory.
+ *
+ * @see <a href="https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html">AWS AppSync Defined Scalars</a>
+ */
+public final class AWSDateTime extends AWSDate {
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTime.java]: No longer dependent on `AWSDate` or `AWSTime`

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTime.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof AWSDateTime)) {
+            return false;
+        }
+
+        AWSDateTime that = (AWSDateTime) obj;
+        return this.getTime() == that.getTime();
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTime.java]: No longer checks instance variable structures

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTimeUtils.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            return parseDateTime(temporal, pos);
+        }
+        if (AWSDate.class.equals(clazz)) {
+            return parseDate(temporal, pos);
+        }
+        if (AWSTime.class.equals(clazz)) {
+            return parseTime(temporal, pos);
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSDateTimeUtils.java]: Removed utils

File: aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSTemporalTypeAdapter.java (hunk 4 above)
Comment (raphkim) [reply aws-api/src/main/java/com/amplifyframework/api/aws/scalar/AWSTemporalTypeAdapter.java]: WHOA this is really cool. I had no idea you could do this.

File: core/src/main/java/com/amplifyframework/core/types/scalar/AWSTemporal.java [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+     * @param string String to verify the presence of a character
+     * @param pos Position to check for presence of a character
+     * @param matchAtLeastOne List of characters to match at least one
+     * @return True if the character at given index matches any one
+     *         of the provided characters
+     */
+    protected static boolean verifyCharAt(
+            @NonNull String string,
+            @NonNull ParsePosition pos,
+            char... matchAtLeastOne
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/core/types/scalar/AWSTemporal.java]: Cool

**Comment by raphkim** [comment]
Closing in favor of #407
